### DISCUSSION TITLE
Configuration fails with VTK for Fedora 26

### DISCUSSION THREAD

**Discussion by CsatiZoltan** [DLR-AMR/t8code#782] [Setup & Installation]
Following the Wiki for the installation, I use the following command:
```bash
[t8code_build]$ ../t8code/configure CFLAGS="-Wall -O0 -g" CXXFLAGS="-Wall -O0 -g" --enable-mpi --enable-static --disable-shared CC=mpicc CXX=mpicxx --prefix=$HOME/t8code_install/ --with-vtk --with-vtk_version_number=9.1 LDFLAGS=-L$HOME/vtk_install/lib64 CPPFLAGS=-I$HOME/vtk_install/include/vtk-9.1
```
which gives the error below:
```
checking for VTK library... checking for VTK Version 9.1... configure: error: Unable to link with vtk library
```
Note that I had to change *vtk_install/lib* to *vtk_install/lib64* on my Fedora Linux as VTK creates a *lib64* directory (might be useful to update the Wiki).

Why does t8code not find my freshly installed VTK?

My *$HOME/vtk_install/* folder has the following tree structure
```
|- lib64
|   |- cmake
|   |   |- vtk-9.1
|- bin
|- include
|   |- vtk-9.1
|- share
```

I could `cd` into all the paths written in the configure command above, so it is not a typo.

**Comment by jmark** [comment]
Could you also upload the `config.log` file? There we might find more information where the problem is.

**Comment by CsatiZoltan** [comment]
Here it is: [config.log](https://github.com/DLR-AMR/t8code/files/12717380/config.log)

**Comment by jmark** [comment]
Can you try it without `--enable-static --disable-shared `?

**Comment by CsatiZoltan** [comment]
I get the same error.

**Comment by jmark** [comment]
The only thing I find in your `config.log` is this:
```
configure:18563: mpicxx -o conftest -Wall -O0 -g -I/home/elsa_home/csati/vtk_install/include/vtk-9.1 -L/home/elsa_home/csati/vtk_install/lib64/cmake conftest.cpp -lgomp -lz -lm  /home/elsa_home/csati/vtk_install/lib64/ >&5
/home/elsa_home/csati/vtk_install/lib64/: file not recognized: Is a directory
```

But I can't diagnose what the problem is from here.

**Comment by CsatiZoltan** [comment]
`cd /home/elsa_home/csati/vtk_install/lib64/` works, so this directory exists. However, the message *file not recognized: Is a directory* implies that it expects a file, not a directory.

Do you have a configuration where you link VTK to t8code? If so, what command do you use for building?

Without VTK, the configuration completes, meaning that this is not a general path issue, but specific to VTK.

If it helps, the VTK CMake files are installed to *$HOME/vtk_install/lib64/cmake/vtk-9.1*.

**Comment by jmark** [comment]
I compiled a setup script which works on my machine.
```bash
#!/bin/bash

git clone 'https://gitlab.kitware.com/vtk/vtk.git'

pushd vtk

  git checkout 'v9.1.0'

popd

mkdir vtk_build

VTK_INSTALL_PREFIX="${PWD}/vtk_install"

pushd vtk_build

  cmake -D CMAKE_INSTALL_PREFIX="${VTK_INSTALL_PREFIX}" -D VTK_USE_MPI=ON ../vtk

  make -j
  make install

popd

export PATH="$PATH:${PWD}/vtk_install/bin"
export LD_LIBRARY_PATH="$LD_LIBRARY_PATH:${PWD}/vtk_install/lib"

git clone 'https://github.com/DLR-AMR/t8code.git'

pushd t8code

  git submodule init
  git submodule update

  ./bootstrap

popd

mkdir t8code_build

T8CODE_INSTALL_PREFIX="${PWD}/t8code_install"
  
pushd t8code_build

  ../t8code/configure \
    --prefix="${T8CODE_INSTALL_PREFIX}" \
    --enable-mpi \
    --with-vtk \
    --with-vtk_version_number=9.1 \
    CFLAGS="-O3" CXXFLAGS="-O3" \
    CC=mpicc CXX=mpicxx \
    LDFLAGS="-L${VTK_INSTALL_PREFIX}/lib" \
    CPPFLAGS="-I${VTK_INSTALL_PREFIX}/include/vtk-9.1"

  make -j
  make install -j

popd
```

Edit: Added backslash after `--prefix="${T8CODE_INSTALL_PREFIX}"`

**Comment by CsatiZoltan** [comment]
Thanks for the script. When I ran it, it installed t8code. However, two strange things happened:
1. Calling the examples with `mpirun` had no effect, as I saw on the VTK output (the .pvtu file included a single .vtu file).
2. When I ran `t8_cmesh_read_from_vtk`, it printed out that this example requires that t8code be linked with VTK.

I think I found the cause of the problem. In your script, there is no backslash after `--prefix="${T8CODE_INSTALL_PREFIX}"`, so all what comes after is neglected: that's why the lacking support for VTK and MPI.

I inserted the backslash and launched the script again. This time, it failed with the same error as I reported yesterday.

It means that at least my installation commands were correct. Could you check what it gives you if you insert the backslash? If it also fails on your side, then I found a bug. If not, I must continue pinpointing the issue on my machine.

**Comment by jmark** [comment]
Silly me. I edited this line after pasting the script to Github. On my machine I ran the script with the backslash. So, yes this script installs t8code linked against VTK 9.1.0. I also tested the exectuables, checked if they are indeed linked against VTK (`ldd ...` command) and checked some of the output.

There must be something different on your machine. In the office I (must) work on (older) Ubuntu systems.

**Comment by jmark** [comment]
I tested above script on my private machine with a quite recent ArchLinux setup. It worked without issues besides applying a small patch https://gitlab.kitware.com/vtk/vtk/-/merge_requests/9857.patch which justs adds a header file for an unknown symbol.

**Comment by jmark** [comment]
@CsatiZoltan You wrote that you are working on a Fedora machine? Which version?

**Comment by CsatiZoltan** [comment]
It's a very old Fedora, I will check its version number later (I'm not in office).

**Comment by CsatiZoltan** [comment]
@jmark I have Fedora26.

**Comment by jmark** [comment]
@CsatiZoltan I spun up a Fedora 26 instance on a VM and tried to setup t8code there. However, the standard packages installed via `dnf` for `gcc` (version 7.3.1) , `cmake` (version 3.11.0), etc. are very old. Fedora 26 is from 2017.

This means one has to install custom versions of the whole compiler chain which is fiddly and hard to provide technical support for. I recommend to upgrade your system.

Feel free to reopen this issue if there are follow-up questions.

**Comment by CsatiZoltan** [comment]
Thanks for trying it out.
With GCC 10.2.0 and CMake 3.25.1, I got the same error.
Then I realized I was stupid: although I changed *lib/* to *lib64/* in `LDFLAGS`, I forgot in the `LD_LIBRARY_PATH` before it.
t8code is now build on my machine!

**Comment by CsatiZoltan** [comment]
This time, I tried to install t8code on our cluster, based on CentOS 7.9 with kernel version 3.10.0-1160.el7.x86_64. I loaded the required modules:
```bash
$ module list
Currently Loaded Modulefiles:
  1) tools/automake/1.16.1
  2) tools/git/2.37.1
  3) tools/cmake/3.23.0
  4) compiler/gcc/9.4.0
  5) mpi/openmpi/4.1.1_gcc94
  6) tools/libtool/2.4.6
```
but the building failed (not a VTK error).

To find the cause of the error, could you tell me what toolchain you use on your clusters? I saw in the [Wiki](https://github.com/DLR-AMR/t8code/wiki/Setting-up-t8code-on-JUWELS-(or-other-Slurm-based-systems)#required-modules) that it is an Intel compiler and MPI. Strangely, on my Fedora desktop, it worked fine with the default GCC 7.3.1.

**Reply by jmark** [reply]
I moved your question to a new discussion thread: https://github.com/orgs/DLR-AMR/discussions/784

**Reply by CsatiZoltan** [reply]
I myself would have opened a new discussion thread if I had had the right.

**Reply by jmark** [reply]
I know. Unfortunately, elevated permissions for Github discussions are only granted to members of our organization. I wish it would be easier and more convenient. I suggest to simply open an issue and if applicable we move it to the discussion section.
